File: python/LeetCode_383_RansomNote.py
# Optimal Solution

"""
def canConstruct(self, ransomNote: str, magazine: str) -> bool:
    for i in set(ransomNote):
        if magazine.count(i) < ransomNote.count(i):
            return False
    else:
        return True
"""


def canConstruct(ransomNote: str, magazine: str) -> bool:
    hash = [0] * 26
    for letter in magazine:
        hash[ord(letter) - 97] += 1

    for letter in ransomNote:
        hash[ord(letter) - 97] -= 1
        if (hash[ord(letter) - 97] < 0): return False

    return True 


canConstruct('aa', 'aab')

File: c/LeetCode_383_RansomNote.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


typedef enum {false, true} bool; 


bool canConstruct(char* ransomNote, char* magazine) {
    int l_ran = strlen(ransomNote);
    int l_mag = strlen(magazine);

    int *hash = (int*) calloc(26, sizeof(int));
    for (int i = 0; i < l_mag; ++i) {
        ++hash[magazine[i] - 'a'];
    }    

    for (int i = 0; i < l_ran; ++i) {
        --hash[ransomNote[i] - 'a'];
        if (hash[ransomNote[i] - 'a'] < 0) return false;
    }

    return true;
}


int main() {

    return 0;
}